File: seo-avengers-2500/quantum-runtime/providers/ibm/ibm-qpu-preflight.mjs
import { spawn } from "node:child_process";
import { createHash } from "node:crypto";
import { mkdir, readFile, writeFile } from "node:fs/promises";
import { dirname, join, resolve } from "node:path";
import { fileURLToPath } from "node:url";

import { compileQaoaExecutableCircuitToOpenQasm3 } from "../../qaoa-openqasm3-compiler.mjs";

const PROVIDER = "IBM_QUANTUM_COMPUTE";
const ENGINE_ID = "NEXUS_IBM_QUANTUM_QPU_PREFLIGHT_V1";
const BRIDGE_ID = "NEXUS_IBM_QUANTUM_QPU_PREFLIGHT_V1";
const DEFAULT_BRIDGE_PATH = fileURLToPath(new URL("./ibm-qpu-preflight.py", import.meta.url));
const MAX_BRIDGE_OUTPUT_BYTES = 64 * 1024 * 1024;
const MAX_TRANSPILER_SEED = 2_147_483_647;
const SHA_RE = /^sha256:[0-9a-f]{64}$/;

function sha256Text(value) {
  if (typeof value !== "string") throw new TypeError("hash input must be string");
  return `sha256:${createHash("sha256").update(Buffer.from(value, "utf8")).digest("hex")}`;
}

function requiredText(value, label) {
  if (typeof value !== "string" || !value.normalize("NFC").trim()) throw new Error(`${label} is required`);
  return value.normalize("NFC").trim();
}

function optionalSecret(value, label) {
  if (value === null || value === undefined || value === "") return null;
  return requiredText(value, label);
}

function optionalTranspilerSeed(value) {
  if (value === null || value === undefined) return null;
  if (!Number.isSafeInteger(value) || value < 0 || value > MAX_TRANSPILER_SEED) {
    throw new Error(`IBM transpilerSeed must be integer between 0 and ${MAX_TRANSPILER_SEED}`);
  }
  return value;
}

function exactKeys(value, expected, label) {
  if (!value || typeof value !== "object" || Array.isArray(value)) throw new Error(`${label} must be object`);
  if (JSON.stringify(Object.keys(value).sort()) !== JSON.stringify([...expected].sort())) throw new Error(`unexpected ${label} keys`);
}

function assertSha(value, label) {
  if (typeof value !== "string" || !SHA_RE.test(value)) throw new Error(`${label} must be sha256 digest`);
  return value;
}

function assertTextDigest(text, digest, label) {
  if (typeof text !== "string") throw new Error(`${label} artifact must be string`);
  if (sha256Text(text) !== assertSha(digest, `${label} digest`)) throw new Error(`${label} artifact digest mismatch`);
}

function sanitizePreflightComponent(response) {
  return sha256Text(JSON.stringify({
    backendDevice: response.preflight.backendDevice,
    logicalQasm3Sha256: response.preflight.logicalQasm3Sha256,
    transpiledCircuitSha256: response.preflight.transpiledCircuitSha256,
    topologySha256: response.preflight.topologySha256,
    capabilitiesSha256: response.preflight.capabilitiesSha256,
    transpilerSeed: response.preflight.transpilerSeed,
  })).slice("sha256:".length);
}

async function writeContentAddressed(path, content) {
  await mkdir(dirname(path), { recursive: true });
  try {
    await writeFile(path, content, { encoding: "utf8", flag: "wx" });
  } catch (error) {
    if (error?.code !== "EEXIST") throw error;
    const existing = await readFile(path, "utf8");
    if (existing !== content) throw new Error(`content-addressed preflight evidence collision:${path}`, { cause: error });
  }
}

function validateResourceEstimate(value) {
  exactKeys(value, ["depth", "gateCount", "twoQubitGateCount"], "IBM preflight resource estimate");
  for (const field of ["depth", "gateCount", "twoQubitGateCount"]) {
    if (!Number.isSafeInteger(value[field]) || value[field] < 0) throw new Error(`IBM preflight ${field} must be non-negative safe integer`);
  }
  return value;
}

export function validateIbmPreflightResponse(response, { logicalCompilation, backendName, transpilerSeed = null }) {
  const expectedSeed = optionalTranspilerSeed(transpilerSeed);
  exactKeys(response, ["artifacts", "bridgeId", "preflight", "schemaVersion"], "IBM preflight response");
  if (response.schemaVersion !== 1 || response.bridgeId !== BRIDGE_ID) throw new Error("unsupported IBM preflight response");

  const preflight = response.preflight;
  exactKeys(preflight, [
    "backendDevice", "backendQubitCount", "capabilitiesSha256", "compiler", "compilerVersion", "logicalQasm3Sha256",
    "logicalQubitCount", "nativeOperationSetSatisfied", "operational", "pendingJobs", "physicalExecutionVerdict", "physicalQpu",
    "provider", "providerSdk", "providerSdkVersion", "qubitCapacitySufficient", "readiness", "reasons", "resourceEstimate",
    "statusMessage", "submissionAttempted", "topologySha256", "transpiledCircuitSha256", "transpilerSeed",
  ], "IBM preflight evidence");
  if (preflight.provider !== PROVIDER || preflight.backendDevice !== backendName) throw new Error("IBM preflight provider/backend identity mismatch");
  if (preflight.physicalQpu !== true) throw new Error("IBM preflight must identify physical QPU");
  if (preflight.submissionAttempted !== false) throw new Error("IBM preflight must never submit a physical job");
  if (preflight.physicalExecutionVerdict !== "NOT_TESTED") throw new Error("IBM preflight cannot promote physical execution verdict");
  if (!Array.isArray(preflight.reasons) || preflight.reasons.some((reason) => typeof reason !== "string" || !reason)) {
    throw new Error("IBM preflight reasons must be string array");
  }
  if (!["READY", "NOT_READY"].includes(preflight.readiness)) throw new Error("IBM preflight readiness invalid");
  for (const field of ["backendQubitCount", "logicalQubitCount", "pendingJobs"]) {
    if (!Number.isSafeInteger(preflight[field]) || preflight[field] < 0) throw new Error(`IBM preflight ${field} invalid`);
  }
  for (const field of ["operational", "qubitCapacitySufficient", "nativeOperationSetSatisfied"]) {
    if (typeof preflight[field] !== "boolean") throw new Error(`IBM preflight ${field} must be boolean`);
  }
  validateResourceEstimate(preflight.resourceEstimate);
  assertSha(preflight.logicalQasm3Sha256, "IBM preflight logical QASM digest");
  assertSha(preflight.transpiledCircuitSha256, "IBM preflight transpiled circuit digest");
  assertSha(preflight.topologySha256, "IBM preflight topology digest");
  assertSha(preflight.capabilitiesSha256, "IBM preflight capabilities digest");
  if (preflight.logicalQasm3Sha256 !== logicalCompilation.qasm3Sha256) throw new Error("IBM preflight logical QASM binding drift");
  const observedSeed = preflight.transpilerSeed ?? null;
  if (expectedSeed === null ? observedSeed !== null : observedSeed !== String(expectedSeed)) {
    throw new Error("IBM preflight transpiler seed binding mismatch");
  }

  const artifacts = response.artifacts;
  exactKeys(artifacts, [
    "capabilitiesJson", "capabilitiesSha256", "logicalQasm3", "logicalQasm3Sha256", "topologyJson", "topologySha256",
    "transpiledQasm3", "transpiledQasm3Sha256",
  ], "IBM preflight artifacts");
  assertTextDigest(artifacts.logicalQasm3, artifacts.logicalQasm3Sha256, "logical QASM3");
  assertTextDigest(artifacts.transpiledQasm3, artifacts.transpiledQasm3Sha256, "transpiled QASM3");
  assertTextDigest(artifacts.topologyJson, artifacts.topologySha256, "backend topology");
  assertTextDigest(artifacts.capabilitiesJson, artifacts.capabilitiesSha256, "backend capabilities");
  if (artifacts.logicalQasm3 !== logicalCompilation.qasm3 || artifacts.logicalQasm3Sha256 !== logicalCompilation.qasm3Sha256) {
    throw new Error("IBM preflight logical circuit artifact drift");
  }
  if (preflight.transpiledCircuitSha256 !== artifacts.transpiledQasm3Sha256
    || preflight.topologySha256 !== artifacts.topologySha256
    || preflight.capabilitiesSha256 !== artifacts.capabilitiesSha256) {
    throw new Error("IBM preflight artifact digest drift");
  }
  if (preflight.readiness === "READY") {
    if (!preflight.operational || !preflight.qubitCapacitySufficient || !preflight.nativeOperationSetSatisfied || preflight.statusMessage.toLowerCase() !== "active" || preflight.reasons.length !== 0) {
      throw new Error("IBM preflight READY contradicts evidence");
    }
  }
  return Object.freeze({ preflight: Object.freeze(preflight), artifacts: Object.freeze(artifacts) });
}

export async function runIbmQpuPreflightBridge({ request, apiKey, instanceCrn, pythonExecutable = "python3", bridgePath = DEFAULT_BRIDGE_PATH, timeoutMillis = 0, spawnImpl = spawn }) {
  const executable = requiredText(pythonExecutable, "pythonExecutable");
  const path = requiredText(bridgePath, "bridgePath");
  const token = requiredText(apiKey, "IBM Quantum API key");
  const crn = requiredText(instanceCrn, "IBM Quantum instance CRN");
  if (!Number.isSafeInteger(timeoutMillis) || timeoutMillis < 0) throw new Error("timeoutMillis must be a non-negative safe integer");
  if (typeof spawnImpl !== "function") throw new Error("spawnImpl must be function");

  return new Promise((resolvePromise, rejectPromise) => {
    const child = spawnImpl(executable, [path], {
      shell: false,
      stdio: ["pipe", "pipe", "pipe"],
      env: {
        ...process.env,
        NEXUS_IBM_QUANTUM_API_KEY: token,
        NEXUS_IBM_QUANTUM_INSTANCE_CRN: crn,
      },
    });
    let stdout = "";
    let stderr = "";
    let settled = false;
    let timer = null;
    const finish = (error, value) => {
      if (settled) return;
      settled = true;
      if (timer) clearTimeout(timer);
      if (error) rejectPromise(error); else resolvePromise(value);
    };
    const append = (current, chunk, label) => {
      const next = current + Buffer.from(chunk).toString("utf8");
      if (Buffer.byteLength(next, "utf8") > MAX_BRIDGE_OUTPUT_BYTES) {
        child.kill("SIGKILL");
        finish(new Error(`IBM preflight bridge ${label} exceeded evidence bound`));
      }
      return next;
    };
    child.on("error", () => finish(new Error("IBM_QPU_PREFLIGHT_PROCESS_START_FAILED")));
    child.stdout.on("data", (chunk) => { stdout = append(stdout, chunk, "stdout"); });
    child.stderr.on("data", (chunk) => { stderr = append(stderr, chunk, "stderr"); });
    child.on("close", (code, signal) => {
      if (settled) return;
      if (code !== 0) {
        finish(new Error(`IBM_QPU_PREFLIGHT_PROCESS_FAILED:${code ?? "null"}:${signal ?? "none"}:${stderr.slice(0, 256)}`));
        return;
      }
      let parsed;
      try { parsed = JSON.parse(stdout); }
      catch { finish(new Error("IBM_QPU_PREFLIGHT_OUTPUT_NOT_JSON")); return; }
      finish(null, parsed);
    });
    if (timeoutMillis > 0) {
      timer = setTimeout(() => {
        child.kill("SIGKILL");
        finish(new Error("IBM_QPU_PREFLIGHT_TIMEOUT"));
      }, timeoutMillis);
      timer.unref?.();
    }
    child.stdin.end(JSON.stringify(request));
  });
}

export function createIbmPreflightFilesystemEvidenceSink({ rootDir }) {
  const root = resolve(requiredText(rootDir, "IBM preflight evidence rootDir"));
  return async ({ preflight, artifacts, request }) => {
    const response = { preflight, artifacts };
    const evidenceDir = join(root, sanitizePreflightComponent(response));
    const files = [
      ["logical.openqasm3", artifacts.logicalQasm3],
      ["transpiled.openqasm3", artifacts.transpiledQasm3],
      ["topology.json", artifacts.topologyJson],
      ["capabilities.json", artifacts.capabilitiesJson],
      ["request.json", `${JSON.stringify(request, null, 2)}\n`],
      ["preflight.json", `${JSON.stringify(preflight, null, 2)}\n`],
    ];
    for (const [name, content] of files) await writeContentAddressed(join(evidenceDir, name), content);
    return Object.freeze({ evidenceDir, preflightSha256: sha256Text(`${JSON.stringify(preflight, null, 2)}\n`) });
  };
}

function buildPreflightRequest({ physicalRequest, backendName, logicalCompilation, transpilerSeed }) {
  return Object.freeze({
    schemaVersion: 1,
    provider: PROVIDER,
    backendName,
    logicalQubitCount: physicalRequest.circuitPayload.logicalQubitCount,
    transpilerSeed,
    logicalCircuitArtifact: Object.freeze({
      format: "OPENQASM_3",
      qasm3: logicalCompilation.qasm3,
      qasm3Sha256: logicalCompilation.qasm3Sha256,
      sourceCircuitSha256: logicalCompilation.sourceCircuitSha256,
    }),
  });
}

export function buildIbmPreflightRequestForContractTest({ physicalRequest, backendName, transpilerSeed = null }) {
  const logicalCompilation = compileQaoaExecutableCircuitToOpenQasm3(physicalRequest.circuitPayload);
  const request = buildPreflightRequest({
    physicalRequest,
    backendName: requiredText(backendName, "IBM Quantum backend name"),
    logicalCompilation,
    transpilerSeed: optionalTranspilerSeed(transpilerSeed),
  });
  return Object.freeze({ request, logicalCompilation });
}

export function createIbmQuantumComputePreflight({
  apiKey = null,
  instanceCrn = null,
  backendName = null,
  transpilerSeed = null,
  evidenceSink = null,
  pythonExecutable = "python3",
  bridgePath = DEFAULT_BRIDGE_PATH,
  timeoutMillis = 0,
  bridgeRunner = runIbmQpuPreflightBridge,
} = {}) {
  const token = optionalSecret(apiKey, "IBM Quantum API key");
  const crn = optionalSecret(instanceCrn, "IBM Quantum instance CRN");
  const backend = optionalSecret(backendName, "IBM Quantum backend name");
  const seed = optionalTranspilerSeed(transpilerSeed);
  const anyConfiguration = token !== null || crn !== null || backend !== null || seed !== null || evidenceSink !== null;
  const fullyConfigured = token !== null && crn !== null && backend !== null;
  if (anyConfiguration && !fullyConfigured) throw new Error("IBM QPU preflight requires apiKey, instanceCrn, and backendName together");
  if (evidenceSink !== null && typeof evidenceSink !== "function") throw new Error("IBM QPU preflight evidenceSink must be function");
  if (typeof bridgeRunner !== "function") throw new Error("IBM QPU preflight bridgeRunner must be function");

  return Object.freeze({
    engineId: ENGINE_ID,
    async run(physicalRequest) {
      if (!fullyConfigured) {
        return Object.freeze({
          schemaVersion: 1,
          engineId: ENGINE_ID,
          provider: PROVIDER,
          verdict: "NOT_TESTED",
          readiness: "NOT_CONFIGURED",
          reasons: Object.freeze(["IBM_QPU_PROVIDER_ACCESS_NOT_CONFIGURED"]),
          physicalExecutionVerdict: "NOT_TESTED",
          submissionAttempted: false,
          preflightEvidence: null,
        });
      }
      const logicalCompilation = compileQaoaExecutableCircuitToOpenQasm3(physicalRequest.circuitPayload);
      const request = buildPreflightRequest({ physicalRequest, backendName: backend, logicalCompilation, transpilerSeed: seed });
      let response;
      try {
        response = await bridgeRunner({ request, apiKey: token, instanceCrn: crn, pythonExecutable, bridgePath, timeoutMillis });
      } catch (error) {
        return Object.freeze({
          schemaVersion: 1,
          engineId: ENGINE_ID,
          provider: PROVIDER,
          verdict: "FAIL",
          readiness: "PREFLIGHT_ERROR",
          reasons: Object.freeze(["IBM_QPU_PREFLIGHT_BRIDGE_FAILED"]),
          physicalExecutionVerdict: "NOT_TESTED",
          submissionAttempted: false,
          preflightEvidence: null,
          errorCode: String(error?.message ?? "IBM_QPU_PREFLIGHT_BRIDGE_FAILED").slice(0, 256),
        });
      }
      const checked = validateIbmPreflightResponse(response, { logicalCompilation, backendName: backend, transpilerSeed: seed });
      if (evidenceSink) await evidenceSink({ preflight: checked.preflight, artifacts: checked.artifacts, request });
      return Object.freeze({
        schemaVersion: 1,
        engineId: ENGINE_ID,
        provider: PROVIDER,
        verdict: checked.preflight.readiness === "READY" ? "PASS" : "INCONCLUSIVE",
        readiness: checked.preflight.readiness,
        reasons: Object.freeze([...checked.preflight.reasons]),
        physicalExecutionVerdict: "NOT_TESTED",
        submissionAttempted: false,
        preflightEvidence: checked.preflight,
      });
    },
  });
}

export const IBM_QPU_PREFLIGHT_ENGINE_ID = ENGINE_ID;
export const IBM_QPU_PREFLIGHT_MAX_TRANSPILER_SEED = MAX_TRANSPILER_SEED;

File: seo-avengers-2500/quantum-runtime/providers/ibm/ibm-qpu-preflight.py
#!/usr/bin/env python3
"""Read-only IBM Quantum Compute preflight for NEXUS Quantum One.

This bridge authenticates, resolves one named backend, validates that it is a
physical QPU, captures backend status/capability/topology evidence, and performs
seeded ISA transpilation of the exact logical circuit. It never submits a
Sampler job and therefore never claims physical execution.
"""

from __future__ import annotations

import hashlib
import json
import os
import sys
import unicodedata
from typing import Any

SCHEMA_VERSION = 1
PROVIDER = "IBM_QUANTUM_COMPUTE"
BRIDGE_ID = "NEXUS_IBM_QUANTUM_QPU_PREFLIGHT_V1"
API_KEY_ENV = "NEXUS_IBM_QUANTUM_API_KEY"
INSTANCE_CRN_ENV = "NEXUS_IBM_QUANTUM_INSTANCE_CRN"
MAX_TRANSPILER_SEED = 2_147_483_647


def _required_text(value: Any, label: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise ValueError(f"{label} is required")
    return unicodedata.normalize("NFC", value.strip())


def _required_int(value: Any, label: str, minimum: int = 0) -> int:
    if not isinstance(value, int) or isinstance(value, bool) or value < minimum:
        raise ValueError(f"{label} must be integer >= {minimum}")
    return value


def _optional_transpiler_seed(value: Any) -> int | None:
    if value is None:
        return None
    seed = _required_int(value, "transpilerSeed", 0)
    if seed > MAX_TRANSPILER_SEED:
        raise ValueError(f"transpilerSeed must be <= {MAX_TRANSPILER_SEED}")
    return seed


def _sha256_text(value: str) -> str:
    return "sha256:" + hashlib.sha256(value.encode("utf-8")).hexdigest()


def _json_safe(value: Any) -> Any:
    if value is None or isinstance(value, (str, bool, int)):
        return value
    if isinstance(value, float):
        return format(value, ".17g")
    if isinstance(value, (list, tuple)):
        return [_json_safe(item) for item in value]
    if isinstance(value, dict):
        return {str(key): _json_safe(item) for key, item in value.items()}
    model_dump = getattr(value, "model_dump", None)
    if callable(model_dump):
        return _json_safe(model_dump())
    to_dict = getattr(value, "to_dict", None)
    if callable(to_dict):
        return _json_safe(to_dict())
    raise TypeError(f"provider value is not JSON-safe:{type(value).__name__}")


def _artifact_json(value: Any) -> str:
    return json.dumps(_json_safe(value), ensure_ascii=False, sort_keys=True, separators=(",", ":"))


def _validate_request(payload: Any) -> dict[str, Any]:
    if not isinstance(payload, dict):
        raise ValueError("preflight request must be object")
    expected = {
        "schemaVersion",
        "provider",
        "backendName",
        "logicalQubitCount",
        "transpilerSeed",
        "logicalCircuitArtifact",
    }
    if set(payload) != expected:
        raise ValueError("unexpected IBM preflight request keys")
    if payload["schemaVersion"] != SCHEMA_VERSION or payload["provider"] != PROVIDER:
        raise ValueError("unsupported IBM preflight schema/provider")
    _required_text(payload["backendName"], "backendName")
    _required_int(payload["logicalQubitCount"], "logicalQubitCount", 1)
    _optional_transpiler_seed(payload["transpilerSeed"])

    artifact = payload["logicalCircuitArtifact"]
    if not isinstance(artifact, dict) or set(artifact) != {
        "format", "qasm3", "qasm3Sha256", "sourceCircuitSha256"
    }:
        raise ValueError("invalid IBM preflight logical circuit artifact")
    if artifact["format"] != "OPENQASM_3":
        raise ValueError("IBM preflight requires OPENQASM_3")
    qasm = _required_text(artifact["qasm3"], "logical QASM3")
    if _sha256_text(qasm) != artifact["qasm3Sha256"]:
        raise ValueError("logical QASM3 digest mismatch")
    for field in ("qasm3Sha256", "sourceCircuitSha256"):
        value = _required_text(artifact[field], field)
        if not value.startswith("sha256:") or len(value) != 71:
            raise ValueError(f"{field} must be sha256 digest")
    return payload


def _backend_is_physical(backend: Any, backend_name: str) -> bool:
    lowered = backend_name.lower()
    if any(token in lowered for token in ("simulator", "fake", "emulator", "statevector")):
        return False
    configuration_fn = getattr(backend, "configuration", None)
    if callable(configuration_fn):
        configuration = configuration_fn()
        if bool(getattr(configuration, "simulator", False)):
            return False
    return True


def _topology_artifact(backend: Any) -> str:
    edges: list[list[int]] = []
    coupling_map = getattr(backend, "coupling_map", None)
    get_edges = getattr(coupling_map, "get_edges", None) if coupling_map is not None else None
    if callable(get_edges):
        edges = sorted([list(map(int, edge)) for edge in get_edges()])
    return _artifact_json({"backend": str(getattr(backend, "name", "")), "edges": edges})


def _capabilities_artifact(backend: Any) -> str:
    target = getattr(backend, "target", None)
    operation_names: list[str] = []
    if target is not None:
        names = getattr(target, "operation_names", None)
        if names is not None:
            operation_names = sorted(str(name) for name in names)
    return _artifact_json({
        "backend": str(getattr(backend, "name", "")),
        "numQubits": int(getattr(backend, "num_qubits", 0) or 0),
        "operationNames": operation_names,
    })


def _transpiler_options(request: dict[str, Any]) -> dict[str, Any]:
    return {
        "optimization_level": 1,
        "seed_transpiler": _optional_transpiler_seed(request["transpilerSeed"]),
    }


def _instruction_names(circuit: Any) -> list[str]:
    names: list[str] = []
    for instruction in getattr(circuit, "data", []):
        operation = getattr(instruction, "operation", None)
        name = getattr(operation, "name", None)
        if name is None and isinstance(instruction, tuple) and instruction:
            name = getattr(instruction[0], "name", None)
        if name is None:
            raise RuntimeError("IBM_PREFLIGHT_TRANSPILED_INSTRUCTION_WITHOUT_NAME")
        names.append(str(name))
    return names


def _two_qubit_gate_count(circuit: Any) -> int:
    count = 0
    for instruction in getattr(circuit, "data", []):
        qubits = getattr(instruction, "qubits", None)
        if qubits is None and isinstance(instruction, tuple) and len(instruction) > 1:
            qubits = instruction[1]
        if qubits is not None and len(qubits) == 2:
            count += 1
    return count


def _readiness(*, operational: bool, status_message: str, capacity_ok: bool, native_ok: bool) -> tuple[str, list[str]]:
    reasons: list[str] = []
    if not operational:
        reasons.append("BACKEND_NOT_OPERATIONAL")
    if status_message.lower() != "active":
        reasons.append("BACKEND_STATUS_NOT_ACTIVE")
    if not capacity_ok:
        reasons.append("LOGICAL_QUBIT_CAPACITY_EXCEEDED")
    if not native_ok:
        reasons.append("TRANSPILED_CIRCUIT_NOT_NATIVE_TO_TARGET")
    return ("READY" if not reasons else "NOT_READY", reasons)


def _execute(request: dict[str, Any]) -> dict[str, Any]:
    api_key = _required_text(os.environ.get(API_KEY_ENV), API_KEY_ENV)
    instance_crn = _required_text(os.environ.get(INSTANCE_CRN_ENV), INSTANCE_CRN_ENV)

    try:
        import qiskit  # type: ignore
        from qiskit import qasm3  # type: ignore
        from qiskit.transpiler import generate_preset_pass_manager  # type: ignore
        import qiskit_ibm_runtime  # type: ignore
        from qiskit_ibm_runtime import IBMQuantumComputeService  # type: ignore
    except Exception as exc:  # pragma: no cover - live provider environment only
        raise RuntimeError("IBM_QPU_PREFLIGHT_DEPENDENCIES_UNAVAILABLE") from exc

    service = IBMQuantumComputeService(
        channel="ibm_quantum_platform",
        token=api_key,
        instance=instance_crn,
    )
    backend = service.backend(request["backendName"])
    backend_name = str(getattr(backend, "name", request["backendName"]))
    if backend_name != request["backendName"]:
        raise RuntimeError("IBM_PREFLIGHT_BACKEND_IDENTITY_MISMATCH")
    if not _backend_is_physical(backend, backend_name):
        raise RuntimeError("IBM_PREFLIGHT_BACKEND_IS_NOT_PHYSICAL_QPU")

    status = backend.status()
    operational = bool(getattr(status, "operational", False))
    pending_jobs = int(getattr(status, "pending_jobs", 0) or 0)
    status_message = str(getattr(status, "status_msg", "") or "")
    backend_qubits = int(getattr(backend, "num_qubits", 0) or 0)
    logical_qubits = request["logicalQubitCount"]
    capacity_ok = backend_qubits >= logical_qubits and backend_qubits > 0

    logical_qasm = request["logicalCircuitArtifact"]["qasm3"]
    circuit = qasm3.loads(logical_qasm)
    if int(circuit.num_qubits) != logical_qubits:
        raise RuntimeError("IBM_PREFLIGHT_LOGICAL_QUBIT_COUNT_MISMATCH")

    pass_manager = generate_preset_pass_manager(backend=backend, **_transpiler_options(request))
    isa_circuit = pass_manager.run(circuit)
    isa_qasm3 = qasm3.dumps(isa_circuit)
    instruction_names = _instruction_names(isa_circuit)
    target = getattr(backend, "target", None)
    target_names = set(str(name) for name in (getattr(target, "operation_names", []) or []))
    native_ok = bool(target_names) and all(name in target_names for name in instruction_names)
    readiness, reasons = _readiness(
        operational=operational,
        status_message=status_message,
        capacity_ok=capacity_ok,
        native_ok=native_ok,
    )

    topology_json = _topology_artifact(backend)
    capabilities_json = _capabilities_artifact(backend)
    transpiled_sha = _sha256_text(isa_qasm3)
    gate_count = int(getattr(isa_circuit, "size")())
    depth = int(getattr(isa_circuit, "depth")())

    return {
        "schemaVersion": SCHEMA_VERSION,
        "bridgeId": BRIDGE_ID,
        "preflight": {
            "provider": PROVIDER,
            "backendDevice": backend_name,
            "physicalQpu": True,
            "operational": operational,
            "statusMessage": status_message,
            "pendingJobs": pending_jobs,
            "logicalQubitCount": logical_qubits,
            "backendQubitCount": backend_qubits,
            "qubitCapacitySufficient": capacity_ok,
            "nativeOperationSetSatisfied": native_ok,
            "transpilerSeed": None if request["transpilerSeed"] is None else str(request["transpilerSeed"]),
            "logicalQasm3Sha256": request["logicalCircuitArtifact"]["qasm3Sha256"],
            "transpiledCircuitSha256": transpiled_sha,
            "topologySha256": _sha256_text(topology_json),
            "capabilitiesSha256": _sha256_text(capabilities_json),
            "resourceEstimate": {
                "gateCount": gate_count,
                "twoQubitGateCount": _two_qubit_gate_count(isa_circuit),
                "depth": depth,
            },
            "providerSdk": "qiskit-ibm-runtime",
            "providerSdkVersion": str(getattr(qiskit_ibm_runtime, "__version__", "unknown")),
            "compiler": "qiskit.generate_preset_pass_manager",
            "compilerVersion": str(getattr(qiskit, "__version__", "unknown")),
            "readiness": readiness,
            "reasons": reasons,
            "submissionAttempted": False,
            "physicalExecutionVerdict": "NOT_TESTED",
        },
        "artifacts": {
            "logicalQasm3": logical_qasm,
            "logicalQasm3Sha256": request["logicalCircuitArtifact"]["qasm3Sha256"],
            "transpiledQasm3": isa_qasm3,
            "transpiledQasm3Sha256": transpiled_sha,
            "topologyJson": topology_json,
            "topologySha256": _sha256_text(topology_json),
            "capabilitiesJson": capabilities_json,
            "capabilitiesSha256": _sha256_text(capabilities_json),
        },
    }


def main() -> int:
    try:
        payload = json.load(sys.stdin)
        request = _validate_request(payload)
        response = _execute(request)
        sys.stdout.write(json.dumps(response, ensure_ascii=False, separators=(",", ":")))
        return 0
    except Exception as exc:
        sys.stderr.write(f"{type(exc).__name__}:{exc}\n")
        return 1


if __name__ == "__main__":
    raise SystemExit(main())
